# Use R/4.3.0
# list all packages needed
packages <- c("RnBeads","RnBeads.mm10","RnBeads.hg19",
              "wateRmelon","RSQLite","utils","lumi","zip",
              "gplots","RColorBrewer","doParallel")
# Bioconductor 3.17 is required for R/4.3.0
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.17")

# Install packages not yet installed
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  BiocManager::install(packages[!installed_packages])
}
invisible(lapply(packages, library, character.only = TRUE))


# Enable parallelization
logger.start(fname=NA)
parallel.isEnabled()
num.cores <-10
parallel.setup(num.cores)

# Set up directories and variables
setwd("C:/Users/jk/Documents/Lab/methylation array/")
data.dir <- "."
idat.dir <- file.path(data.dir, "205832350006")

# Two annotation for two separate datasets: DMSO only (B and NB) OR full (B, NB; DMSO, GSK343)
# Sample sheets are generated by the core service provider
sample.annotation <- file.path(data.dir, "PAS23023_Mo-Methy_sample sheet_DMSO.csv")
sample.table <- read.delim(file="PAS23023_Mo-Methy_sample sheet_DMSO.csv",sep = ",") # This is not needed to run the code.
sample.annotation.full <- file.path(data.dir, "PAS23023_Mo-Methy_sample sheet.csv")
analysis.dir <- "./RnBeads"
report.dir <- file.path(analysis.dir, "reports")
data.source <- c(idat.dir,sample.annotation)
data.source.full <- c(idat.dir,sample.annotation.full)

# Import RepeatMasker table
# Chr, start, end, repClass
tab.rmsk <- read.delim("./mm10_rmskClass.bed",
                       header=T,
                       stringsAsFactors=F)
colnames(tab.rmsk) <- c("Chromosome","Start","End","name")
rnb.set.annotation("rmsk", tab.rmsk,  description="rmsk",assembly="mm10")

# set option to include RepeatMasker annotation in addition to defaults
rnb.options(region.types=c("tiling","genes","promoters","cpgislands","rmsk"),
            assembly="mm10")

# Deprecated: save annotation for future reference
# annotation.filename <- file.path(analysis.dir,
#                                  "custom_anno.RData")
# rnb.save.annotation(annotation.filename, "rmsk",assembly="mm10")
# 
# rnb.load.annotation(annotation.filename, "rmsk")
# rnb.region.types()


# make RnBeads S4 object
# "result" directory must NOT be there
# there are two data sets: .full includes all samples to generate heatmaps at the end. But I want to limit my initialy comparison to DMSO B vs. DMSO NB cells. Thus, I create "result" object separately.
result <- rnb.run.import(data.source=data.source,
                         data.type="infinium.idat.dir",
                         dir.reports=file.path(analysis.dir, "result"))
result.full  <- rnb.run.import(data.source=data.source.full,
                               data.type="infinium.idat.dir",
                               dir.reports=file.path(analysis.dir, "resultFull"))
# preprocessing that includes both filtering and normalization
rnb.set <- result$rnb.set
rnb.set.unfiltered <- rnb.set
result.preprocess <- rnb.run.preprocessing(rnb.set.unfiltered,
                                           dir.reports=file.path(analysis.dir, "result"))
rnb.set <- result.preprocess$rnb.set

# full dataset preprocessing
rnb.set.full <- result.full$rnb.set
rnb.set.full.unfiltered <- rnb.set.full
result.full.preprocess <- rnb.run.preprocessing(rnb.set.full.unfiltered,
                                           dir.reports=file.path(analysis.dir, "resultFull"))
rnb.set.full <- result.full.preprocess$rnb.set

# low dimensional representation NOT CRUCIAL

dred.sites <- rnb.execute.dreduction(rnb.set)
dred.anno <- rnb.execute.dreduction(rnb.set, target="mm_anno")
dred <- list(sites=dred.sites,anno=dred.anno)
sample.colors <- ifelse(pheno(rnb.set)$Sample_Type=="B","red","blue")
plot(dred[[1]]$mds$manhattan, col=sample.colors)
plot(dred.sites[[1]]$euclidean, col=sample.colors)

# low dimensional representation for full data set

dred.sites.full <- rnb.execute.dreduction(rnb.set.full)
dred.anno.full <- rnb.execute.dreduction(rnb.set.full, target="mm_anno")
dred.full <- list(sites=dred.sites.full, anno=dred.anno.full)


## differential analysis

cmp.cols <- "Sample_Type"

# get.table() returns analysis statistics for each differential probe, but not probe annotation
# annotation() returns annotation information for each differential probe

# genes
diffmeth.genes <- rnb.execute.computeDiffMeth(rnb.set, cmp.cols, region.types="genes")
comparison.genes <- get.comparisons(diffmeth.genes)[1]
tab.genes <- get.table(diffmeth.genes,comparison.genes, region.type="genes",return.data.frame=T)
aa.genes <- annotation(rnb.set,type="genes") # returns information about given "type"
# promoters
diffmeth.promoters <- rnb.execute.computeDiffMeth(rnb.set, cmp.cols, region.types="promoters")
comparison.promoters <- get.comparisons(diffmeth.promoters)[1]
tab.promoters <- get.table(diffmeth.promoters,comparison.promoters, region.type="promoters",return.data.frame=T)
aa.promoters <- annotation(rnb.set,type="promoters")
# cpgislands
diffmeth.cpgislands <- rnb.execute.computeDiffMeth(rnb.set, cmp.cols, region.types="cpgislands")
comparison.cpgislands <- get.comparisons(diffmeth.cpgislands)[1]
tab.cpgislands <- get.table(diffmeth.cpgislands,comparison.cpgislands, region.type="cpgislands",return.data.frame=T)
aa.cpgislands <- annotation(rnb.set,type="cpgislands")
# RepeatMasker
diffmeth.rmsk <- rnb.execute.computeDiffMeth(rnb.set, cmp.cols, region.types="rmsk")
comparison.rmsk <- get.comparisons(diffmeth.rmsk)[1]
tab.rmsk <- get.table(diffmeth.rmsk,comparison.rmsk, region.type="rmsk",return.data.frame=T)
aa.rmsk <- annotation(rnb.set,type="rmsk")

# tab rownames(integer) are the same as meth(rnb.set) rownames, NOT annotation rownames
annotated.tab.genes <- data.frame(tab.genes,aa.genes, row.names=NULL)
annotated.tab.cpgislands <- data.frame(tab.cpgislands,aa.cpgislands, row.names=NULL)
annotated.tab.promoters <- data.frame(tab.promoters,aa.promoters, row.names=NULL)
annotated.tab.rmsk <- data.frame(tab.rmsk,aa.rmsk, row.names=NULL)


# the rownames of the .order dataframe refers to the row numbers of top (combinedRank) differential probes in the annotated.tab.sites data frame, which is not ordered.
# this occurs as the default
# for example, order(annotated.tab.genes[,"combinedRank"]) returns row numbers starting with the smallest combinedRank value to the largest
annotated.tab.genes.order <- annotated.tab.genes[order(annotated.tab.genes[,"combinedRank"]),]
annotated.tab.promoters.order <- annotated.tab.promoters[order(annotated.tab.promoters[,"combinedRank"]),]
annotated.tab.cpgislands.order <- annotated.tab.cpgislands[order(annotated.tab.cpgislands[,"combinedRank"]),]
annotated.tab.rmsk.order <- annotated.tab.rmsk[order(annotated.tab.rmsk[,"combinedRank"]),]

# histogram of delta beta values for rmsk...
hist(annotated.tab.rmsk.order[1:200,3],
     breaks=20,
     xlim=c(-0.6,0.5),
     yaxt="none",xaxt="none",
     ylab=NA, xlab=NA,
     main=NA,
     )
axis(2,las=2,font=2,cex.axis=2.5)
axis(1,las=2,font=2,cex.axis=2.5)


# frequency of each functional feature
# annotated.tab.sites.order1000<- annotated.tab.sites.order$name[1:1000]
# table(annotated.tab.sites.order1000)
# frequency in the original bed file
# dist <- tab.anno$name
# table(dist)
# crucial that type="rmsk" is set for correct rowname values
topDiffBeta.gene.full <- meth(rnb.set.full, type="genes")[as.integer(rownames(annotated.tab.genes.order)),]
topDiffBeta.promoters.full <- meth(rnb.set.full, type="promoters")[as.integer(rownames(annotated.tab.promoters.order)),]
topDiffBeta.cpgislands.full <- meth(rnb.set.full, type="cpgislands")[as.integer(rownames(annotated.tab.cpgislands.order)),]
topDiffBeta.rmsk.full <- meth(rnb.set.full, type="rmsk")[as.integer(rownames(annotated.tab.rmsk.order)),]

# heatmap
colfunc<-colorRampPalette(c("blue","black","red"))


pdf(file="beta_heatmap_top200_genes.pdf")
par(font=2, font.axis=2, font.lab=2,cex.lab=1.5, cex.axis=1.5,
    lty=1)
heatmap.2(topDiffBeta.gene.full[1:200,c(1,5,9,2,6,10,3,7,11,4,8,12)],
          col=colfunc(10), scale="none", Rowv=F, Colv=T,
          cexCol=1, labCol=NA, labRow=NA,srtCol=0, adjCol=c(0.5,0),
          density.info = "none",trace="none", dendrogram = "column",
          symkey=FALSE,symbreaks=F,revC = FALSE, RowSideColors = rep("white",200),
          breaks=c(0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1),
          lmat=rbind(c(5,0,4),c(3,1,2)),
          lhei=c(1,4),
          lwid=c(1.75,0.1, 3.5),margins=c(5,12),cexRow=1.2)
dev.off()


pdf(file="beta_heatmap_top200_promoters.pdf")
par(font=2, font.axis=2, font.lab=2,cex.lab=1.5, cex.axis=1.5,
    lty=1)
heatmap.2(topDiffBeta.promoters.full[1:200,c(1,5,9,2,6,10,3,7,11,4,8,12)],
          col=colfunc(10), scale="none", Rowv=F, Colv=T,
          cexCol=1, labCol=NA, labRow=NA,srtCol=0, adjCol=c(0.5,0),
          density.info = "none",trace="none", dendrogram = "column",
          symkey=FALSE,symbreaks=F,revC = FALSE, RowSideColors = rep("white",200),
          breaks=c(0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1),
          lmat=rbind(c(5,0,4),c(3,1,2)),
          lhei=c(1,4),
          lwid=c(1.75,0.1, 3.5),margins=c(5,12),cexRow=1.2)
dev.off()

pdf(file="beta_heatmap_top200_cpgislands.pdf")
par(font=2, font.axis=2, font.lab=2,cex.lab=1.5, cex.axis=1.5,
    lty=1)
heatmap.2(topDiffBeta.cpgislands.full[1:200,c(1,5,9,2,6,10,3,7,11,4,8,12)],
          col=colfunc(10), scale="none", Rowv=F, Colv=T,
          cexCol=1, labCol=NA, labRow=NA,srtCol=0, adjCol=c(0.5,0),
          density.info = "none",trace="none", dendrogram = "column",
          symkey=FALSE,symbreaks=F,revC = FALSE, RowSideColors = rep("white",200),
          breaks=c(0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1),
          lmat=rbind(c(5,0,4),c(3,1,2)),
          lhei=c(1,4),
          lwid=c(1.75,0.1, 3.5),margins=c(5,12),cexRow=1.2)
dev.off()

pdf(file="beta_heatmap_top200_rmsk.pdf")
par(font=2, font.axis=2, font.lab=2,cex.lab=1.5, cex.axis=1.5,
    lty=1)
colSide <- annotated.tab.rmsk.order$name

colSide <-gsub("LTR","slateblue3",colSide)
colSide <- gsub("LINE","green3",colSide)
colSide <- gsub("SINE","green3",colSide)
colSide <- gsub("DNA","black",colSide)
colSide <- gsub("Simple_repeat","yellow3",colSide)
colSide <- gsub("Other","blue",colSide)
colSide <- gsub("Unknown","blue",colSide)

heatmap.2(topDiffBeta.rmsk.full[1:200,c(1,5,9,2,6,10,3,7,11,4,8,12)],
          col=colfunc(10), scale="none", Rowv=F, Colv=T,
          cexCol=1, labCol=NA, labRow=NA,srtCol=0, adjCol=c(0.5,0),
          density.info = "none",trace="none", dendrogram = "column",
          symkey=FALSE,symbreaks=F,revC = FALSE, RowSideColors = colSide[1:200],
          breaks=c(0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1),
          lmat=rbind(c(5,0,4),c(3,1,2)),
          lhei=c(1,4),
          lwid=c(1.75,0.3, 3.5),margins=c(5,12),cexRow=1.2)
dev.off()


# Getting tables required for GEO submission
meth <- M(rnb.set.full, row.names=T)
unmeth <- U(rnb.set.full, row.names=T)
beta <- meth(rnb.set.full, row.names=T)
pval <- dpval(rnb.set.full, row.names=T)

signal.table <-data.frame(
                unmeth[,1], meth[,1], pval[,1],
                unmeth[,2], meth[,2], pval[,2],
                unmeth[,3], meth[,3], pval[,3],
                unmeth[,4], meth[,4], pval[,4],
                unmeth[,5], meth[,5], pval[,5],
                unmeth[,6], meth[,6], pval[,6],
                unmeth[,7], meth[,7], pval[,7],
                unmeth[,8], meth[,8], pval[,8],
                unmeth[,9], meth[,9], pval[,9],
                unmeth[,10], meth[,10], pval[,10],
                unmeth[,11], meth[,11], pval[,11],
                unmeth[,12], meth[,12], pval[,12]
)

beta.table <-data.frame(
          beta[,1],pval[,1],
          beta[,2],pval[,2],
          beta[,3],pval[,3],
          beta[,4],pval[,4],
          beta[,5],pval[,5],
          beta[,6],pval[,6],
          beta[,7],pval[,7],
          beta[,8],pval[,8],
          beta[,9],pval[,9],
          beta[,11],pval[,10],
          beta[,11],pval[,11],
          beta[,12],pval[,12])
          
write.table(signal.table, "./signal_table.xls", sep="\t", row.names=T,
            quote=F)
write.table(beta.table, "./beta_table.xls", sep="\t", row.names=T,
            quote=F)
